_docs/migration_guides/upgrading_to_terragrunt_0.19.x.md

```diff
@@ -0,0 +1,268 @@
+
+  extra_arguments "custom_vars" {
+    commands  = ["apply", "plan"]
+    arguments = ["-var", "foo=42"]
+  }
+}
+
+remote_state {
+  backend = "s3"
+  config = {
+    bucket         = "my-terraform-state"
+    key            = "${path_relative_to_include()}/terraform.tfstate"
+    region         = "us-east-1"
+    encrypt        = true
+    dynamodb_table = "my-lock-table"
+  }   
+}  
+```
+
+
+### Move input variables into inputs
+
+When we were using `terraform.tfvars` files for Terragrunt configuration, we were piggybacking on the fact that 
+Terraform [automatically loads variables from tfvars 
+files](https://www.terraform.io/docs/configuration/variables.html#variable-definitions-tfvars-files) to set variables
+for our modules:
+
+```hcl
+# terraform.tfvars
+
+# Terragrunt configuration
+terragrunt = {
+  terraform {
+    # ...
+  }
+  
+  remote_state {
+    # ...
+  }
+}
+
+# Input variables to set for your Terraform module
+instance_type  = "t2.micro"
+instance_count = 10
+``` 
+
+With the move to `terragrunt.hcl`, we no longer get this behavior for free. However, Terragrunt can simulate this 
+behavior for you if you define your input variables by specifying `inputs = { ... }`:
+
+```hcl
+# terragrunt.hcl
+
+terraform {
+  # ...
+}
+
+remote_state {
+  # ...  
+}
+
+# Input variables to set for your Terraform module
+inputs = {
+  instance_type  = "t2.micro"
+  instance_count = 10
+}  
+``` 
+
+Whenever you run a Terragrunt command, such as `terragrunt apply`, Terragrunt will make these variables available to
+your Terraform module as environment variables.
+
+### Use first-class expressions
+
+Terraform 0.11 only allowed special behavior, such as function calls, using "interpolation syntax," where you wrapped
+the code with `${...}`. Terragrunt included a handful of functions you could call using interpolation syntax, but 
+_only_ within the `terragrunt = { ... }` block:
+
+```hcl
+# terraform.tfvars
+
+terragrunt = {
+  terraform {
+    extra_arguments "retry_lock" {
+      # Using a function within interpolation syntax
+      commands  = "${get_terraform_commands_that_need_locking()}"
+      arguments = ["-lock-timeout=20m"]
+    }
+  }
+}
+
+# Using interpolation syntax outside of the terragrunt config did NOT work before
+foo = "${get_env("FOO", "default")}"
+```
+
+Terraform 0.12 has moved to HCL2, which has first-class support for expressions. That means you can call functions 
+without having to wrap them in `${...}`. Terragrunt embraces HCL2, and thanks to HCL2's nice parser, that means we not
+only get first-class expressions, but we can also use those expressions _everywhere_ in `terragrunt.hcl`!
+
+```hcl
+# terragrunt.hcl
+
+terraform {
+  extra_arguments "retry_lock" {
+    # Using a function within first-class expressions!
+    commands  = get_terraform_commands_that_need_locking()
+    arguments = ["-lock-timeout=20m"]
+  }
+}
+
+# This now works with Terragrunt 0.19.x and newer!
+foo = get_env("FOO", "default")
```

Comment:
Should this be wrapped in `inputs`?

Reply:
Yes, fixed! https://github.com/gruntwork-io/terragrunt/pull/731/commits/47391c61b67de90d19323c340980b5d32c353d39